### ISSUE TITLE
Python 3 migration plan

### ISSUE THREAD

**Issue by robnagler** [radiasoft/sirepo#1560]
This task is to create a plan for the Python 2 (py2) to Python 3 (py3)  migration. It depends greatly on #1234.

**Comment by robnagler** [comment]
---------- Forwarded message ---------
From: Rollin Thomas <>
Date: Fri, Sep 13, 2019 at 9:20 AM
Subject: [Users] Python 2 End-of-Life in 2020 and Consequences at NERSC
To: <>


Dear NERSC User,

As you may know, Python 2 will reach its end of life on January 1, 2020.
<https://devguide.python.org/#status-of-python-branches> After its last
release, Python 2 will cease to exist as an active project: No development,
no bug fixes, no patches, etc. *This is important because users must
actively transition to Python 3, which is not backward-compatible with
Python 2.*

Developers of many packages including NumPy, SciPy, Matplotlib, Pandas, and
Scikit-Learn have pledged to drop support for Python 2 “no later than 2020.”
<https://python3statement.org/> We can expect support for Python 2 from
these developers to continue to wither away over the next year.

This has some implications. Even if NERSC committed to support Python 2
users in perpetuity, our ability to help users with Python 2 issues would
be curtailed by the lack of support from the developer community.

While Python 2 will not spontaneously stop working on January 1 of next
year, users should expect changes in support at NERSC and elsewhere. Here
is our plan for the retirement of Python 2:

   - Cori’s default Python module, python/2.7-anaconda-2019.07, remains
   default until the allocation year (AY) rollover in early January 2020.
   - At the AY rollover, we will change the default Python module to one
   based on a Python 3 distribution.
   - The old Python 2 module will not be deleted. It will remain available
   for use, but users must specify the version suffix explicitly.
   - No new installations of Python 2 packages or modules will happen after
   the 2020 AY rollover.
   - During the next Cori OS upgrade (possible sometime in 2020), the
   existing Python 2 module will be retired. NERSC will actively support only
   Python 3 on Perlmutter and future systems.

Of course, users will be able to create Python 2 conda environments and use
them so long as that is viable. You may already have noticed deprecation
warnings from your Python applications’ outputs; do not ignore these
warnings! You should develop a migration plan soon if you don’t have one
already.

At the Python 3 Statement website (linked above) there are a few links
under the “Why?” section that may be helpful to you in preparing your
migration plan and easing the transition from Python 2 to Python 3. These
seem especially helpful:

https://docs.python.org/3/howto/pyporting.html
https://python-3-for-scientists.readthedocs.io/en/latest/

If you have any questions, please let us know via a ticket at help.nersc.gov
.

**Comment by robnagler** [comment]
In order to switch to py3 for Sirepo, we need to either run all codes in Docker, which is in progress.

However, this begs the question what are the py2-only codes that are directly imported in `sirepo.template.*`. Here are the codes with question marks:

* openPMD-viewer
* rslinac
* rswarp
* synergia

We don't import `warp` directly in the server, just `openpmd_viewer`. Other codes like SRW, SDDS, etc. are already being used in py3 so I think we can safely assume that they run in py3. Many other more popular libraries are used (numpy, scikit, etc.), which shouldn't cause concern.

We can drop rslinac, since it's not a supported code. When it is again, we can fix it.

We can fix rswarp. I'm guessing `OpenPMD-viewer` works, because people are using Warp in py3.

That leaves synergia, which doesn't seem to support py3 afaict. @bruhwiler wants to import MAD-X files directly so maybe we can decouple from synergia's MAD-X import. That would leave calculate_bunch_parameters, which uses some stuff, but that could be a separate process call (that's how the job code does it). I'm guessing it could be decoupled. pconstants could be denormalized. Similarly, save_report_data is going to be a separate process. 

The reason I bring this up is that Dec would be a good month to do a one time py2-py3 conversion. If we get the job support released, or even if we don't, we can use job_process. Everything would be broken until we got a complete py3 migration. Nobody is importing sirepo directly. Pykern is already pretty much py3 compatible. 

@moellep thoughts?

**Comment by robnagler** [comment]
See https://github.com/radiasoft/sirepo/issues/1848

**Comment by robnagler** [comment]
Stuff that can be done today:

* replace `xrange` with `range`

**Comment by robnagler** [comment]
Need https://github.com/radiasoft/pykern/issues/61

**Comment by robnagler** [comment]
We've had a lot of success with py3, and we can run codes as py3 or py2 so I'm pushing to switch the server to py3. Working on updating beamsim now:

*  #2338
* https://github.com/radiasoft/download/issues/83 elegant
* https://github.com/radiasoft/download/issues/78 drop fpm (side issue but related)
* https://github.com/radiasoft/download/issues/84 boost
* https://github.com/radiasoft/download/issues/73 warp

**Comment by robnagler** [comment]
Just a quick update. synergia, warppba and warpvnd seem to be working. There are data issues with jspec and elegant. Some tests are failing, and yet to be debugged. The jspec issue is unrelated to py3, but the fact that it is happening on my VM (and not other VMs we've tried) is concerning. The same with the elegant data difference (animation_test is failing).

**Comment by robnagler** [comment]
@ahebnl We are going to be switching master to python 3 only today. 

When you merge master into your development branch, you'll need to update your VM.

**Comment by robnagler** [comment]
warpvnd has problems with raw data file on some reports:

- [ ] Source Tab, Field Comparison Report, "Raw Data File" broken
- [ ] Source Tab, Field Report, "Raw Data File" broken
- [ ] Visualization Tab, Field Animation "Raw Data File" broken
- [ ] Visualization Tab, Current Animation "Raw Data File" broken
- [ ] Visualization Tab, Impact Density "Raw Data File" broken
- [ ] "Lattice Gate 3D" example, Visualization Particle Trace 3d "Impact Density" broken

**Comment by robnagler** [comment]
@moellep The other ones are fixed.

This is in rswarp, which I'll attempt to fix:

```txt
Traceback (most recent call last):
  File "parameters.py", line 544, in <module>
    min_s, max_s, d = plot_density.generate_plot_data_for_faces_3d(c)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/rswarp/diagnostics/ImpactDensity.py", line 161, in generate_plot_data_for_faces_3d
    for face in cond.generate_faces_3d():
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/rswarp/diagnostics/ConductorTemplates.py", line 170, in generate_faces_3d
    self.cbounds[(axis + 1) % 3 + 3], xn),
  File "<__array_function__ internals>", line 6, in linspace
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/numpy/core/function_base.py", line 121, in linspace
    .format(type(num)))
TypeError: object of type <class 'numpy.float64'> cannot be safely interpreted as an integer.
```

I'm seeing this in the logs, but I don't think it has to do with impact density:

```txt
../sirepo/server.py:137:api_errorLogging 10.10.10.1: javascript error: {
    "cause": "<no cause>",
    "message": "TypeError: Cannot read property 'screenDim' of undefined",
    "stackTrace": "TypeError: Cannot read property 'screenDim' of undefined
    at refreshAxes (http://v.radia.run:8000/static/js/warpvnd.js?20200417.183111:4680:50)
    at refresh (http://v.radia.run:8000/static/js/warpvnd.js?20200417.183111:4577:17)
    at m.$scope.resize (http://v.radia.run:8000/static/js/warpvnd.js?20200417.183111:4930:17)
    at http://v.radia.run:8000/static/js/sirepo-plotting.js?20200417.183111:728:23
    at m.$broadcast (http://v.radia.run:8000/static/js/ext/angular.min.js?20200417.183111:157:98)
    at http://v.radia.run:8000/static/js/sirepo.js?20200417.183111:1114:20
    at later (http://v.radia.run:8000/static/js/sirepo-components.js?20200417.183111:3739:29)
    at m (http://v.radia.run:8000/static/js/ext/angular.min.js?20200417.183111:117:319)
    at m.$digest (http://v.radia.run:8000/static/js/ext/angular.min.js?20200417.183111:151:185)
    at m.$apply (http://v.radia.run:8000/static/js/ext/angular.min.js?20200417.183111:155:35)",
    "type": "clientException",
    "url": "http://v.radia.run:8000/warpvnd#/visualization/oUHhIwsM"
}
```

This is the line:

```txt
                    var screenDim = axisCfg[dim].screenDim;
```

**Comment by e-carlin** [comment]
[Fixed pkdc on bytes object and using pkio.read_text](https://github.com/radiasoft/sirepo/commit/75a568892ae9509f2ddf272cdeff142ef5d17e36)

**Comment by robnagler** [comment]
rscode-rsbeams is rebuilt so if you are already py3, you can just `sudo dnf update -y rscode-rsbeams`.

**Comment by robnagler** [comment]
ec2b1e30 fixes friction force

**Comment by robnagler** [comment]
synergia
Import an elegant lattice (.lte) file is broken.

zgoubi
"GSI KAOS Spectrometer with TOSCA" on Lattice tab select the edit button for the DIPOLE element causes server error.

zgoubi.dat import broken. On the Visualization page for any example, run the simulation and then select download "zgoubi.dat" and \
then import that file.

**Comment by robnagler** [comment]
These are fixed.

I'm going to merge master now.

**Comment by robnagler** [comment]
Fix 9eab965e
